File: 1406-stone-game-iii/1406-stone-game-iii.cpp
class Solution {
public:
    string stoneGameIII(vector<int>& stoneValue) {
        int res = 0;
        vector<int> dp(stoneValue.size(), INT_MIN);
        res = solve(stoneValue, dp, 0);
        if (res > 0) {
            return "Alice";
        } else if (res < 0) {
            return "Bob";
        }
        return "Tie";
    }

private:
    int solve(vector<int>& stoneValue, vector<int>& dp, int i) {
        if (i == stoneValue.size()) {
            return 0;
        }
        if (dp[i] == INT_MIN) {
            dp[i] = max(dp[i], stoneValue[i] - solve(stoneValue, dp, i + 1));
            if (i + 1 < stoneValue.size()) {
                dp[i] = max(dp[i], stoneValue[i] + stoneValue[i + 1] -
                                   solve(stoneValue, dp, i + 2));
            }
            if (i + 2 < stoneValue.size()) {
                dp[i] =
                    max(dp[i], stoneValue[i] + stoneValue[i + 1] +
                                 stoneValue[i + 2] - solve(stoneValue, dp, i + 3));
            }
        }
        return dp[i];
    }
};

File: 1406-stone-game-iii/1406-stone-game-iii.py
class Solution:
    def stoneGameIII(self, stoneValue: List[int]) -> str:
        n = len(stoneValue)
        dp = [None for _ in range(n)]
        def solve(i: int) -> int:
            if i == n:
                return 0
            if dp[i] == None:
                dp[i] = float('-inf')
                dp[i] = max(dp[i], stoneValue[i] - solve(i + 1))
                if i + 1 < n:
                    dp[i] = max(dp[i], stoneValue[i] + stoneValue[i + 1] - solve(i + 2))
                if i + 2 < n:
                    dp[i] = max(dp[i], stoneValue[i] + stoneValue[i + 1] + stoneValue[i + 2] - solve(i + 3))
            return dp[i]
        res = solve(0)
        if res > 0:
            return "Alice"
        elif res < 0:
            return "Bob"
        return "Tie"
